language/types_values_variables.md

```diff
@@ -456,6 +461,333 @@ For conversion from `String` both float and integer formats are supported.
 * In `String` format, integer prefixes for hex and binary radix are understood (but not octal since
   floating point in string format may start with a '0').
 
+### Timespan ([from, to])
+
+Since version 4.8.0
+
+Represents a range of timespan values. The default is the range +/- Infinity.
+
+A timespan is an duration, measured in seconds, with nano-second precision.
+
+A `Timespan` range behaves as an `Integer` range and accepts integer, float, or timespan values when
+specifying the range. It is however, not possible to iterate over a `Timespan` range.
+
+#### Type Algebra on Timespan
+
+    Timespan ∪ Timespan            → Timespan
+    Timespan ∪ Integer             → Numeric
+    Timespan ∪ Numeric             → Numeric
+    Timespan ∪ (T ∈ Scalar)        → Scalar
+    Timespan ∪ (T ∉ Scalar)        → Any
+    Timespan[a, b] ∪ Timespan[c, d]   → Timespan[min(a, c), max(b,d)]
+
+#### Timespan.new
+
+A new `Timespan` can be created from `Integer`, `Float`, `String`, and `Hash` values. Several variants of the constructor are provided.
+
+##### Timespan from seconds
+
+When a Float is used, the decimal part represents fractions of a second.
+
+```puppet
+function Timespan.new(
+  Variant[Float, Integer] $value
+)
+```
+
+##### Timespan from days, hours, mintues, seconds, and fractions of a second
+
+The arguments can be passed separately in which case the first four, days, hours, minutes, and seconds are mandatory and the rest are optional.
+All values may overflow and/or be negative. The internal 128-bit nano-second integer is calculated as:
+
+```
+(((((days * 24 + hours) * 60 + minutes) * 60 + seconds) * 1000 + milli_seconds) * 1000 + micro_seconds) * 1000 + nano_seconds
+```
+
+```puppet
+function Timespan.new(
+  Integer $days, Integer $hours, Integer $minutes, Integer $seconds,
+  Integer $milli_seconds = 0, Integer $micro_seconds = 0, Integer $nano_seconds = 0
+)
+```
+
+or, all arguments can be passed as a `Hash`, in which case all entries are optional:
+
+```puppet
+function Timespan.new(
+  Struct[{
+    Optional[negative] => Boolean,
+    Optional[days] => Integer,
+    Optional[hours] => Integer,
+    Optional[minutes] => Integer,
+    Optional[seconds] => Integer,
+    Optional[milli_seconds] => Integer,
+    Optional[micro_seconds] => Integer,
+    Optional[nano_seconds] => Integer
+  }] $hash
+)
+```
+
+##### Timespan from String and format directive patterns
+
+The first argument is parsed using the format optionally passed as a string or array of strings. When an array is used, an attempt
+will be made to parse the string using the first entry and then with each entry in succession until parsing succeeds. If the second
+argument is omitted, an array of default formats will be used.
+
+An exception is raised when no format was able to parse the given string.
```

Comment:
We do not raise exception in puppet - the concept of exception does not exist. We do "raise errors". (If other text in the spec talks about exceptions it should be changed). Or, since this is a specification and the mechanism is not specified - it can simply say "It is an error if no format was able to..."